### ISSUE TITLE
Common process methods

### ISSUE THREAD

**Issue by bhaugen** [valueflows/valueflows#264]
A question came up when Lynn and I were discussing recipes about process classifications. I said a lot of common processes exist that are independent of specific recipes, but have their own descriptions and techniques that are common to all of the recipes they might be used in. And I promised to look them up. 

This is similar to taxonomies of resource classifications, as in agrovoc.

Here are some lists of common processes, methods, and techniques:
* [in manufacturing](https://en.wikipedia.org/wiki/List_of_manufacturing_processes)
* [cooking](https://en.wikipedia.org/wiki/List_of_cooking_techniques)
* [circular economy](https://github.com/valueflows/valueflows/issues/96)

Depending on your familiarity with the methods specified in a particular recipe, it might be good to have a reference to an extended description of how to do it.

**Comment by fosterlynn** [comment]
So are those ProcessClassifications?  Or something else that maybe relates to ProcessClassification or to RecipeProcess (or whatever its final name is)?  Perhaps the model is a bit more complex than what we have?

**Comment by elf-pavlik** [comment]
I don't think we try to decide just on name ProcessClassification vs. ProcessRecipe || ProcessBlueprint || ProcessTemplate etc. I see those two (sides of vs.) as two distinct cases.

https://en.wikipedia.org/wiki/List_of_cooking_techniques seems like possibly useful set of taxonomy references (ProcessClassification) any Process and (ProcessRecipe || ProcessBlueprint || ProcessTemplate) could reference those classifications.

Classification (Taxonomy) layer != Recipe layer

**Comment by fosterlynn** [comment]
>https://en.wikipedia.org/wiki/List_of_cooking_techniques seems like possibly useful set of taxonomy references (ProcessClassification) any Process and (ProcessRecipe || ProcessBlueprint || ProcessTemplate) could reference those classifications.

The problem I was seeing with the cooking techniques is that you might use more than one inside one process.  I don't see that people should make their processes smaller than they should be (governed by resource flows, scheduling, etc.) just because there is a technique described by itself.  And vice versa, of course, M:M.  But I don't know how complex we want to make all of that....

**Comment by elf-pavlik** [comment]
> The problem I was seeing with the cooking techniques is that you might use more than one inside one process.

If we don't think one should reference all of those classifications with *vf:processClassifiedAs*, we could add something like *vf:includesStepsClasifiedAs*.

I still think we still haven't address how to compose processes together in different ways, and that one 'bigger process' can get broken down in multiple 'smaller processes' in many different ways, each of which in the end will have same set of events as the 'bigger process' plus/minus some 'service', 'issue' & 'receive' events caused by different process orchestration and inventory management approaches.

For recipes which in practice require many different processes, agents may often choose to delegate some of those processes to other agents, which in turn can aggregate them in bigger processes. Sometimes relying on some stock (resource) for bridging, sometimes on *service* events. In that case we will also get all kinds of agreements *under* which those events and prior commitments will happen. To get the *service* part down IMO we need to work more with *work* and *use* since they often will come either as direct input or as input to another process which then provides a service.

**Comment by bhaugen** [comment]
> we could add something like vf:includesStepsClasifiedAs.

> I still think we still haven't address how to compose processes together in different ways, and that one 'bigger process' can get broken down in multiple 'smaller processes' in many different ways, each of which in the end will have same set of events as the 'bigger process' plus/minus some 'service', 'issue' & 'receive' events caused by different process orchestration and inventory management approaches.

I expect people will want "steps" inside of processes. Bil McCarthy likes steps, and will include them in his next monograph, which he has been working on for at least the last year. I have personally been resistant because I need to code this stuff, and follow the resource and process flows in every direction. 

If "steps" were like processes within processes, and followed the same IPO configuration, that would keep the code simpler than if they were yet another concept and artifact that would also be an exception to the flow logic. Don't underestimate the effects of exceptions to a common logic.

That being said, I won't be surprised if they end up in the mix. DocuBricks, which I am looking at now, and think we will need to support, has both steps and a few other new concepts and artifacts that we will either need to work with or translate to.

**Comment by elf-pavlik** [comment]
> If "steps" were like processes within processes, and followed the same IPO configuration, that would keep the code simpler than if they were yet another concept and artifact that would also be an exception to the flow logic. Don't underestimate the effects of exceptions to a common logic.

Since we make 'Flow Event', 'Flow Commitment', 'Flow Intent', (Flow Recipe?) the primary set of constructs, and 'Resource', 'Process', 'Agreement' etc. in a way secondary. We could see if all those 'Flow *' when referencing multiple instances of 'Process' and multiple instances of 'Agreement' would give enough information to reason about how those referenced processes and agreements 'overlap' / 'complement' each other.

I would also stay cautious if those REA steps only focus on manufacturing products or also take into account providing services with *work* and *use* handled as well as *produce* / *consume* and *issue* / *receive*.

**Comment by fosterlynn** [comment]
>I would also stay cautious if those REA steps only focus on manufacturing products or also take into account providing services with work and use handled as well as produce / consume and issue / receive.

I don't really see that it makes any difference in this case, although I could be missing something.  You could have any of the same inputs for service outputs; and might need "steps" for either one.

Re. steps:  I would lean towards exploring parent/child processes.  This would handle also the case of a large project-size process where people could log "overhead" work.

**Comment by bhaugen** [comment]
> where people could log "overhead" work.

Like for example cleaning, which gets neglected in some orgs we know...

**Comment by elf-pavlik** [comment]
>> Re. steps: I would lean towards exploring parent/child processes. This would handle also the case of a large project-size process where people could log "overhead" work

> Like for example cleaning, which gets neglected in some orgs we know..

:+1:
